packages/datadog-shimmer/src/shimmer.js

```diff
@@ -141,52 +136,111 @@ function safeWrapper (original, wrapper) {
-  const innerWrapped = wrapper(function (...args) {
-    // We need to stash the callState here because of recursion.
-    const callState = currentCallState
-    callState.startCall()
-    const retVal = original.apply(this, args)
-    if (isPromise(retVal)) {
-      retVal.then(callState.endCall.bind(callState))
-    } else {
-      callState.endCall(retVal)
+  let innerWrapped
+  // Fast path for async methods. Those are always returning promises, so need
+  // to check for that. That also prevents the need to change the prototype later.
+  if (isAsyncFunction(original)) {
+    innerWrapped = wrapper(function (...args) {
+      // This is only used in safe mode. It's a simple state machine to track if the
+      // original method was called and if it returned. We need this to determine if
+      // an error was thrown by the original method, or by us. We'll use one of these
+      // per call to a wrapped method.
+      // We need to stash the callState here because of recursion.
+      const callState = currentCallState
+      callState[0] = IN_PROGRESS
+      return original.apply(this, args).then((value) => {
+        callState[0] = COMPLETED
+        callState[1] = value
+        return value
+      })
+    })
+  } else {
+    innerWrapped = wrapper(function (...args) {
+      // This is only used in safe mode. It's a simple state machine to track if the
+      // original method was called and if it returned. We need this to determine if
+      // an error was thrown by the original method, or by us. We'll use one of these
+      // per call to a wrapped method.
+      // We need to stash the callState here because of recursion.
+      const callState = currentCallState
+      callState[0] = IN_PROGRESS
+      const retVal = original.apply(this, args)
+      if (isPromise(retVal)) {
+        return retVal.then((value) => {
+          callState[0] = COMPLETED
+          callState[1] = value
+          return value
+        })
+      }
+      callState[0] = COMPLETED
+      callState[1] = retVal
+      return retVal
+    })
+  }
+
+  // The wrapped function is the one that will be called by the user.
+  // It calls our version of the original function, which manages the
+  // callState. That way when we use the errorHandler, it can tell where
+  // the error originated.
+  if (isAsyncFunction(original)) {
+    // Fast path for async methods. Those are always returning promises, so need
+    // to check for that. That also prevents the need to change the prototype later.
+    // TODO(BridgeAR): Check if the overhead with async hooks and the async methods
+    // is higher than the overhead of the prototype change. That way it's possible
+    // to check for the async methods in copyProperties and skip the prototype change.
+    return function (...args) {
+      currentCallState = [NOT_STARTED, undefined]
+      const callState = currentCallState
+
+      return innerWrapped.apply(this, args).catch((error) => {
```

Comment:
I would add a comment about this error handler being a copy of `handleError` and that they probably should be kept in sync. If you ever edit one, you should probably edit the other.

Reply:
Obsolete due to the removal of safe mode